### ISSUE TITLE
Migration issue from conan 1.3 to 1.11

### ISSUE THREAD

**Issue by fourbft** [conan-io/conan#4386]
We have a problem migrating from conan 1.3 to 1.11.
Some of our conan packages have overloaded the ConanFile initializer function `__init__` in the conanfile.py like this:

    from conans import CMake
    from ConanTemplate import ConanTemplate
    ...
    class ConanComponent( ConanTemplate.ConanGeneralComponent):`
        name = "SomeComponent"
        ...

        def __init__( self, output, runner, settings, conanfile_directory, user = None, channel = None):
            ConanTemplate.ConanGeneralComponent.__init__( self, output, runner, settings, conanfile_directory, user, channel)
            self.cmake = CMake( self, parallel = True)
    ...

The first issue with this kind of conanfile.py is, as you will have guessed, the number of arguments to the `__init__` function which has obviously changed from 6 to only 4 somewhere on the way from conan version 1.3 to 1.11.
This results in an error when trying to build the package and even when trying to install a prebuilt (with conan 1.3) package with conan 1.11:

`ERROR: <path_to_conan_cache>/export/conanfile.py: __init__() takes at most 5 arguments (7 given)`

Luckily all of our affected conan packages derive from an intermediate ConanTemplate.ConanGeneralComponent class which then derives from the ConanFile class. So, what we tried to solve the problem was to add an `__init__` initializer function to the ConanGeneralComponent class (which did not have one, so far) that does the conversion of the number of arguments like this:

    class ConanGeneralComponent(ConanFile):
        ...
        def __init__( self, *args):
            ConanFile.__init__( self, *(args[:4]))
        ...

This seems to work. So when rolling out the new conan version 1.11, we would also roll out a new version of the ConanTemplate.ConanGeneralComponent class and that way, we could build and install old packages (with the 6 argument `__init__` function) with the new conan version, that only supplies 4 arguments (and only accepts 4 arguments in the ConanFile class).
I know, this is not great, but it seems to work.

The bigger problem is, actually, that the instantiation of the CMake object on the second line of the `__init__` function doesn't work anymore either. The error message is:

    ERROR: <path_to_conan_cache>/export/conanfile.py: 'tuple' object has no attribute 'get_save'

As far as I have analyzed the problem, the reason seems to be, that inside the `__init__` function, the self.settings attribute still is a simple tuple, while `CMake( self, parallel = True)` wants to access self.settings as an instance of type `Settings` (it tries to call `get_save` on it). In later parts of the conanfile.py (e. g. inside the build() function), self.settings is actually of type Settings.

In conan 1.3, the conanfile.py files as shown above worked flawlessly. In conan 1.11, they don't anymore. We've learned, that it is probably not such a good idea to overwrite the `__init__` function in a conanfile.py and we have already removed these everywhere. But the problem that currently keeps us from upgrading conan is, that with the new conan version, we could no longer even install prebuilt binary packages of the older versions of the packages from our conan server.

So the question is: Does anyone have any ideas, how to get around this problem? Is there something we could do in the ConanTemplate.ConanGeneralComponent __init__ function to convert the self.settings tuple into a Settings class instance to make the CMake( self) constructor work again? Any other ideas?

Of course, we could keep an old conan version around and use it for installing old, prebuilt conan packages. But that's not so great, really.

To help us debug your issue please explain:

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [X] I've specified the Conan version, operating system version and any tool that can be relevant.
- [X] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by fourbft** [comment]
Hi @memsharded 

yes, we've learned the lesson to not mess with internals. :-) 
Only that at the time, we were not aware, that the constructor was considered an internal, but then you are of course right, that it was not officially documented either. The reason we put the instantiation of the CMake object in the initializer was just to avoid duplicated code: we needed the CMake object both in the build() and in the package() methods and did not want to duplicate the instantiation and all the `cmake.definitions[...]` settings. That's why we thought a single class attribute would be a good idea.

Anyway, we've already changed all affected packages. So the current versions of the packages don't have this problem anymore. 

The only issue we have is, that there are now many old binary versions of the packages (some 70 affected packages with about a dozen versions each) uploaded on the conan server and still in use by several projects, that will no longer even be installable when we upgrade to the current conan version. So the main question this issue is all about is: Do you have any idea, what we could temporarily do (maybe in our intermediate ConanTemplate class) to be able to use the latest conan version and still be able to install those 'defect' old packages (rebuilding the old versions would not be required). This would also only need to be an interim solution, until those old prebuilt packages are no longer needed (so for a year or so). And of course we would also implement a huge warning message, if such a construct (own `__init__` function - with or without CMake instantiation in it) should show up in any package again.

The only solution that we can think of is to keep the old conan version around and use both versions in parallel. But that might lead to numerous other problems. If there is no feasible solution, the best thing for us would probably be to update all projects to the latest versions of all packages. Doable, but quite a lot of work.

**Comment by memsharded** [comment]
> The reason we put the instantiation of the CMake object in the initializer was just to avoid duplicated code: we needed the CMake object both in the build() and in the package() methods and did not want to duplicate the instantiation and all the cmake.definitions[...] settings. That's why we thought a single class attribute would be a good idea.

Yes, to avoid that duplication, you could be using a helper method, check for example this one: https://docs.conan.io/en/latest/reference/build_helpers/cmake.html#example The key, is that it is not a class or object attribute, but just a local variable, which creation is encapsulated in a method. In general, having class or object attributes in Conan is not a good idea, because the state might not be stored from call to call.

> Anyway, we've already changed all affected packages. So the current versions of the packages don't have this problem anymore.

Great, many thanks!

> The only issue we have is, that there are now many old binary versions of the packages (some 70 affected packages with about a dozen versions each) uploaded on the conan server and still in use by several projects, that will no longer even be installable when we upgrade to the current conan version. 

Actually, I think that you might be ok, you can change the recipe, put it into the cache with ``export`` (not ``create``), and I don't see anything there that will cause a change in the computed package binary IDs, so your existing binaries should be ok, even if the recipe changed. The only thing is when you do a ``conan search <ref>`` it will list the binaries as "outdated from recipe" (or yellow in the table of ``conan search <ref> --table=file.html``. 

You could try downloading some packages and uploading them to a temporary repo, then testing against that repo. Please try if possible and tell me. Thanks!

**Comment by memsharded** [comment]
Hi @fourbft 

Is the migration issue not an issue anymore, because you already changed the recipes? Could we close this issue? Thanks!

**Comment by fourbft** [comment]
Hi @memsharded,
well, it's still an issue, but we have worked around it. We fixed all conanfile.py's to not define their own __init__ function (we are now following your helper method pattern). And for the projects, that will still use old, affected packages for a while, we have set up a separate compile machine (virtual) that still uses conan 1.3.
So, yes, you can close this issue. Thanks for your support!

**Comment by memsharded** [comment]
Ok, sorry to hear that you need to do that setup, but yes, as there is little that we could do here, we can close the issue then. Thanks for all the feedback!
